layer4/routes.go

```diff
@@ -127,7 +127,8 @@ func wrapRoute(route *Route, logger *zap.Logger) Middleware {
 			// route must match at least one of the matcher sets
 			matched, err := route.matcherSets.AnyMatch(cx)
 			if err != nil {
-				logger.Error("matching connection", zap.Error(err))
+				logger.Error("matching connection", zap.String("remote", cx.RemoteAddr().String()), zap.Error(err))
+				return nil // return nil so the error does not get logged again
```

Comment:
Where would it be logged again? I don't see anywhere else we use the value `err`.

Reply:
It would be logged again in [server.go#L109](https://github.com/mholt/caddy-l4/blob/acf0ea80461d5931e927fc1e9b3d186655a62417/layer4/server.go#L109)

Reply:
Interesting, are you seeing that in your logs? How does `err` get returned from here?

Reply:
The return statement is completely new to fix this:
> The first bug (at least I think it is one 😄) is that when a matcher returns an error the processing is not stopped. Instead the next matcher is tried and if that matches its handlers are executed even if the connection is broken.

You can see example logs of the processing not stopping in the initial post. When using `return err` instead one would see the double logging.

Reply:
I follow so far, except I don't see where we are using `return err` anywhere. We always return `nextCopy.Handle(cx)` instead.

(Sorry if I'm being a bit slow to understand.)

Reply:
I think we are talking past each other :)

I included the comment to signify we are returning `nil` on purpose, because otherwise it may look like a bug.  

You are right we mostly return `nextCopy.Handle(cx)` except for exactly this case, since we want to stop the execution of the chain. But we already logged (handled) the error, so it is wrong to return the error (like it's often the case in `if err != nil` blocks). Because the last return value from the chain will eventually appear in [server.go#L106](https://github.com/mholt/caddy-l4/blob/acf0ea80461d5931e927fc1e9b3d186655a62417/layer4/server.go#L106) and thus would get logged in [server.go#L109](https://github.com/mholt/caddy-l4/blob/acf0ea80461d5931e927fc1e9b3d186655a62417/layer4/server.go#L109) which we do not want because we already logged it.

Reply:
Hi @mholt, did you found the time to think about this?

Reply:
@ydylla I'm so sorry for the delay on this PR. (It's not fair to you, that's for sure.) I was traveling this last week and also taking a little break from sprinting on Caddy stuff after the 2.6 release (to avoid burnout), and now I'm way behind on notifications. :sweat_smile: I'll be catching up soon!

If we can find another maintainer who is willing to review and approve this and other PRs, that might be a good idea so it's not always blocked by me. I feel bad about that! I might look around and see who would be willing to accept the invitation. This plugin just isn't nearly as active as the main Caddy project so it has fewer candidates.

Reply:
No problem at all, I understand that you are very busy. Don't feel bad, you are doing good work with caddy.  
But you have to endure me pining you from time to time 😄 

I just used the delay to clean up the commits and fix the failing tests in the l4socks module.